NeoML/src/Dnn/DnnReferenceRegister.cpp

```diff
@@ -20,47 +20,73 @@ limitations under the License.
 
 namespace NeoML {
 
-CDnnReferenceRegister::CDnnReferenceRegister( CDnn* _originalDnn ) :
-	learningState( false ),
-	referenceCounter( -1 ),
-	originalDnn( _originalDnn ),
-	originalRandom( new CRandom( _originalDnn->Random() ) )
+CDnnReferenceRegister::CDnnReferenceRegister( IMathEngine& mathEngine, CDnn& dnn ) :
+	CDnnReferenceRegister( new CReferenceDnnInfo( dnn.Random(), nullptr ), mathEngine )
 {
-	NeoAssert( _originalDnn != nullptr );
-	if( originalDnn->referenceDnnRegister.referenceCounter == 0 ) {
-		originalDnn->referenceDnnRegister.learningState = originalDnn->IsLearningEnabled();
-	}
-	originalDnn->referenceDnnRegister.referenceCounter += 1;
+	// Copy dnn using serialization to get the new dnn of necessary life time
+	CMemoryFile file;
+	CArchive archive( &file, CArchive::store );
+	dnn.Serialize( archive );
+	archive.Close();
+	file.SeekToBegin();
+
+	// Allow to Serialize() this time
+	CReferenceDnnInfo* tmp = nullptr;
+	swap( tmp, originalDnn.referenceDnnInfo );
```

Comment:
To avoid such juggling, make a private method serialize() in CDnn without any checks and call it here. Public Serialize will be just call to serialize() preceded with assert (!isReferenceDnn) .

Reply:
Serialize is **virtual recursive** method of `CBaseLayer`.
It is the most simple implementation, that I found.